runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -1157,16 +1721,43 @@ MM_SchedulingDelegate::updateGMPStats(MM_EnvironmentVLHGC *env)
 
-	U_64 incrementalScanTime = (U_64) (((double) j9time_hires_delta(0, incrementalMarkStats->getScanTime(), J9PORT_TIME_DELTA_IN_MICROSECONDS)) / ((double) _extensions->gcThreadCount));
-	UDATA concurrentBytesScanned = concurrentMarkStats->_bytesScanned;
+	uint64_t incrementalScanTime = (uint64_t) (((double) j9time_hires_delta(0, incrementalMarkStats->getScanTime(), J9PORT_TIME_DELTA_IN_MICROSECONDS)) / ((double) _extensions->gcThreadCount));
+	uintptr_t concurrentBytesScanned = concurrentMarkStats->_bytesScanned;
 
-	_historicTotalIncrementalScanTimePerGMP = (U_64) ((_historicTotalIncrementalScanTimePerGMP * incrementalScanTimePerGMPHistoricWeight) + (incrementalScanTime * (1 - incrementalScanTimePerGMPHistoricWeight)));
-	_historicBytesScannedConcurrentlyPerGMP = (UDATA) ((_historicBytesScannedConcurrentlyPerGMP * bytesScannedConcurrentlyPerGMPHistoricWeight) + (concurrentBytesScanned * (1 - bytesScannedConcurrentlyPerGMPHistoricWeight)));
+	_historicTotalIncrementalScanTimePerGMP = (uint64_t) ((_historicTotalIncrementalScanTimePerGMP * incrementalScanTimePerGMPHistoricWeight) + (incrementalScanTime * (1 - incrementalScanTimePerGMPHistoricWeight)));
+	_historicBytesScannedConcurrentlyPerGMP = (uintptr_t) ((_historicBytesScannedConcurrentlyPerGMP * bytesScannedConcurrentlyPerGMPHistoricWeight) + (concurrentBytesScanned * (1 - bytesScannedConcurrentlyPerGMPHistoricWeight)));
 
 	Trc_MM_SchedulingDelegate_updateGMPStats(env->getLanguageVMThread(), _historicTotalIncrementalScanTimePerGMP, incrementalScanTime, _historicBytesScannedConcurrentlyPerGMP, concurrentBytesScanned);
 }
 
-U_64
+void 
+MM_SchedulingDelegate::updatePgcTimePrediction(MM_EnvironmentVLHGC *env)
+{
+	/* 
+	 * Create a model that passes through (minimumEdenRegions,minimumPgcTime) and (current eden size in regions, pgcTime) 
+	 * By remembering historic values of _pgcTimeIncreasePerEdenFactor, it is possible to reasonably accuratly predict how long PGC will take, if eden were to change size.
+	 * 
+	 * Note: The formula derived for predicting pgc times, require that x1 and x2 are in GB.
+	 */
+	double x1 = (double)_regionManager->getRegionSize() / 1000000000;
+	double y1 = (double)minimumPgcTime;
+
+	double x2 = (double)getCurrentEdenSizeInBytes(env) / 1000000000;
+	double y2 = (double)_historicalPartialGCTime;
+
+	/* 
+	 * Calculate how closely related PGC is to eden time. The closer _pgcTimeIncreasePerEdenFactor is to 1.0, the more directly changing eden size will impact pgc time.
+	 * The higher _pgcTimeIncreasePerEdenFactor is from 1, the less changing eden size will affect pgc time. 
+	 * In certain edge cases where eden is very small (minimumEdenRegions in size), or pgc time is very small, skip this set of calculation, since the results will not be correct
+	 */
+	if ((x1 < x2) && (y1 < y2)) {
+		double timeDiff = y1 - y2;
+		double edenSizeRatio = ((x1 + 1.0) / (x2 + 1.0));
+		_pgcTimeIncreasePerEdenFactor = pow(edenSizeRatio, (1.0/timeDiff));
```

Comment:
We should report this value in a trace point. Perhaps you can piggyback on an existing one at caller site?

Reply:
Agreed that this should be reported. 
As for which tracepoint to use, I don't see any tracepoints at caller site right now. I'll take a look at some of the other variables that get set at end of PGC, and include any metrics that currently aren't being reporting into this new tracepoint. If all variables are being reported, I'll just add a tracepoint in this function, and be a bit more verbose (ie, include `x1`, `x2`, `timeDiff`, etc)

Reply:
```suggestion
		_pgcTimeIncreasePerEdenFactor = pow(edenSizeRatio, (1.0 / timeDiff));
```